CSS @imports in HTML missing quote marks are mistakenly hidden from the Preload Scanner

Ahmad-S792 opened the pull request:
#### 6851ca14bf62fb07839b4416333db9d5af417f38
<pre>
CSS @imports in HTML missing quote marks are mistakenly hidden from the Preload Scanner

<a href="https://bugs.webkit.org/show_bug.cgi?id=191466">https://bugs.webkit.org/show_bug.cgi?id=191466</a>

Reviewed by Chris Dumez.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Partial Merge - <a href="https://chromium.googlesource.com/chromium/src.git/+/30ecdaf63da0c5e1d81caf93c9836ec5564d9647">https://chromium.googlesource.com/chromium/src.git/+/30ecdaf63da0c5e1d81caf93c9836ec5564d9647</a>

This fixes a couple of bugs with the CSSPreloadScanner that made it miss
rules in which the URL wasn&apos;t quoted or the rule didn&apos;t end with a
semicolon etc.

* Source/WebCore/html/parser/CSSPreloadScanner.cpp:
(parseCSSStringOrURL): Add comments and also add &quot;if&quot; clause for &apos;single&apos; and
&apos;double&apos; clauses and do other clean-ups
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-quote.tentative.html: Add Test Case
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-quote.tentative-expected.txt: Add Test Case Expectation
* LayoutTests/imported/w3c/web-platform-tests/loading/resources/dummy.css: Add Test Resource
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/loading/preloader-css-import-no-quote.tentative-expected.txt: Add Platform Specific Expectation

Canonical link: <a href="https://commits.webkit.org/261254@main">https://commits.webkit.org/261254@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/111405bae166e1165481c3ca5eb027986328bf9d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109652 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18772 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42397 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1133 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118775 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113548 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20239 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9971 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/619 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115405 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15063 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98295 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101921 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97037 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29947 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43281 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11503 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31289 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85062 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12159 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8229 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17525 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50898 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7804 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13903 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rniwa reviewed (dismissed):
We need to add tests.

anttijk commented on Source/WebCore/html/parser/CSSPreloadScanner.cpp:
```diff
@@ -129,8 +129,12 @@ inline void CSSPreloadScanner::tokenize(UChar c)
             m_ruleValue.append(c);
+            // When reading the rule and hitting ')', which signifies the URL end, emit the rule.
+            if (c == ')')
+                emitRule();            
```
I think this is wrong since there may be more things to parse after the first `)` like in

`url("foo.css") layer(foo)`

This is causing the test failure.

Ahmad-S792 replied:
I found this commit post this to add layer support in CSSPreload and also removing this makes us only pass one test out of three.

Ahmad-S792 replied:
I tried multiple cases in local testing but I think it might be something needs more time. Meanwhile removing this still progresses one test without any regression, so I think we can settle for it time being. So I pushed this commit. :-)

cdumez reviewed (approved):
LGTM

Ahmad-S792 committed:
CSS @imports in HTML missing quote marks are mistakenly hidden from the Preload Scanner

https://bugs.webkit.org/show_bug.cgi?id=191466

Reviewed by Chris Dumez.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Partial Merge - https://chromium.googlesource.com/chromium/src.git/+/30ecdaf63da0c5e1d81caf93c9836ec5564d9647

This fixes a couple of bugs with the CSSPreloadScanner that made it miss
rules in which the URL wasn't quoted or the rule didn't end with a
semicolon etc.

* Source/WebCore/html/parser/CSSPreloadScanner.cpp:
(parseCSSStringOrURL): Add comments and also add "if" clause for 'single' and
'double' clauses and do other clean-ups
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-quote.tentative.html: Add Test Case
* LayoutTests/imported/w3c/web-platform-tests/loading/preloader-css-import-no-quote.tentative-expected.txt: Add Test Case Expectation
* LayoutTests/imported/w3c/web-platform-tests/loading/resources/dummy.css: Add Test Resource
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/loading/preloader-css-import-no-quote.tentative-expected.txt: Add Platform Specific Expectation

Canonical link: https://commits.webkit.org/261254@main
